src/main/scala/org/renci/cam/QueryService.scala

```diff
@@ -186,66 +89,113 @@ object QueryService extends LazyLogging {
           queryGraph.nodes.map(a => s"?${a.id}_type") :::
           queryGraph.edges.map(a => s"?${a.id}")
       selectClause = s"SELECT DISTINCT ${ids.mkString(" ")} "
-      valuesClause = {
-        var bindings = mutable.ListBuffer[String]()
-        bindings += sparqlLines.mkString("")
-        instanceVarsToTypes.foreach(a =>
-          a.foreach(b =>
-            nodeTypes.get(b._2) match {
-              case Some(v) => bindings += s"?${b._1} rdf:type $v ."
-              case None => bindings += ""
-            }))
-        s"${bindings.mkString("\n")}"
-      }
+      moreLines = for {
+        (subj, typ) <- instanceVarsToTypes.flatten
+        v <- nodeTypes.get(typ)
+      } yield s"?$subj rdf:type $v ."
+      valuesClause = (sparqlLines ++ moreLines).mkString("\n")
       limitSparql = if (limit > 0) s" LIMIT $limit" else ""
+      prefixMap <- readPrefixes
+      prefixes = prefixMap.map(entry => s"PREFIX ${entry._1}: <${entry._2}>").mkString("\n")
       queryString = s"$prefixes\n$selectClause\n$whereClause\n$valuesClause \n } $limitSparql"
       query <- Task.effect(QueryFactory.create(queryString))
       response <- SPARQLQueryExecutor.runSelectQuery(query)
     } yield response
-  }
 
-  def parseResultSet(queryGraph: TranslatorQueryGraph, resultSet: ResultSet): RIO[ZConfig[AppConfig], TranslatorMessage] =
+  def parseResultSet(queryGraph: TRAPIQueryGraph, resultSet: ResultSet): RIO[ZConfig[AppConfig], TRAPIMessage] =
```

Comment:
I would like to see if we can break this method up into much smaller pieces. Also, it will be simpler to avoid using mutable ListBuffers.

Reply:
I removed all the ListBuffers.

Reply:
Not letting you take all the credit, I removed 2 of the 4.  Still, nice work.